packages/headless/src/features/sort-criteria/structured-sort-slice.ts

```diff
@@ -0,0 +1,27 @@
+import {createReducer} from '@reduxjs/toolkit';
+import {change} from '../history/history-actions';
+import {
+  registerSortCriterion,
+  updateSortCriterion,
+} from './sort-criteria-actions';
+import {restoreSearchParameters} from '../search-parameters/search-parameter-actions';
+
+import {getStructuredSortInitialState} from './structured-sort-state';
+import {parseCriterionExpression} from './criteria-parser';
+
+export const structuredSortReducer = createReducer(
+  getStructuredSortInitialState(),
+  (builder) => {
+    builder
+      .addCase(registerSortCriterion, (_, action) => action.payload)
+      .addCase(updateSortCriterion, (_, action) => action.payload)
+      .addCase(change.fulfilled, (state, action) => {
+        return action.payload?.sort ?? state;
+      })
+      .addCase(restoreSearchParameters, (state, action) => {
+        return action.payload?.sortCriteria
+          ? parseCriterionExpression(action.payload.sortCriteria)
```

Comment:
Now that I have slept on this, I guess I could have just used `sortCriteria` and this parser directly in the action instead of creating a whole new reducer and part of the state

Reply:
Can you expand on this ? 

Because, yeah, If you can just piggy back on the existing sort slice, then that's great, because it means less code to main, and one less "public" thing in headless..

Once a new slice has been added, remember that we can not remove it (unless doing major version changes), and it becomes part of the public API of headless. The shape of the state is part of the public interface of the library.

Could you try the idea out and see if it removes a lot of code ?

Reply:
I agree, I think parsing the string is the better way to go based on the last update.

When we initially discussed this feature, the PL api only supported certain sort criteria.
Using the existing actions/controller would have been misleading to devs, and someone would eventually report a bug where they try to set a sort, but it does not work.

If the plan is for the PL api to eventually support all sort criteria offered for the search use-case (cc: @acote-coveo), then it would be possible to reuse the existing actions and state, greatly simplifying the amount of code.

Reply:
:wave: Picked this up from @francoislg . Here's what it looks like when using the `sortCriteria` directly: https://github.com/coveo/ui-kit/pull/1171/files

It removes the structured sort state & slice.

Reply:
It looks good 👍

Reply:
Merged here!